cmd/common/snapshot/process_test.go

```diff
@@ -0,0 +1,152 @@
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/snapshot/process/types"
+)
+
+type stsTestCase struct {
+	val      uint64
+	valMax   uint64
+	width    uint64
+	expected string
+}
+
+func TestCreateTree(t *testing.T) {
+	t.Parallel()
+
+	events := []*types.Event{
+		{
+			Pid: 42,
+		},
+		{
+			Pid:       43,
+			ParentPid: 42,
+		},
+		{
+			Pid:       44,
+			ParentPid: 42,
+		},
+		{
+			Pid:       45,
+			ParentPid: 42,
+		},
+		{
+			Pid:       46,
+			ParentPid: 43,
+		},
+	}
+
+	tree, err := createTree(events)
+	if err != nil {
+		t.Fatalf("fail to create tree: %v", err)
+	}
+
+	expectedPid := 42
+	if tree.event.Pid != expectedPid {
+		t.Fatalf("root PID is wrong, expected %d got %d", expectedPid, tree.event.Pid)
+	}
+
+	childrenNumber := len(tree.children)
+	expectedChildrenNumber := 3
+	if childrenNumber != expectedChildrenNumber {
+		t.Fatalf("expected %d children got %d", expectedChildrenNumber, childrenNumber)
+	}
+
+	expectedChildrenPIDs := []int{43, 44, 45}
+parentLoop:
+	for _, node := range tree.children {
+		for _, expectedChildrenPID := range expectedChildrenPIDs {
+			if node.event.Pid == expectedChildrenPID {
+				continue parentLoop
+			}
+		}
+		t.Fatalf("node PID is wrong, expected one from %v got %d", expectedChildrenPIDs, node.event.Pid)
+	}
+
+	var specificNode *nodeProcess
+	searchingPid := 43
+	for _, node := range tree.children {
+		if node.event.Pid == searchingPid {
+			specificNode = node
+
+			break
+		}
+	}
+	if specificNode == nil {
+		t.Fatalf("no node with PID %d", searchingPid)
+	}
+
+	childrenNumber = len(specificNode.children)
+	expectedChildrenNumber = 1
+	if childrenNumber != expectedChildrenNumber {
+		t.Fatalf("expected %d children got %d", expectedChildrenNumber, childrenNumber)
+	}
+
+	pid := specificNode.children[0].event.Pid
+	expectedChildPid := 46
+	if pid != expectedChildPid {
+		t.Fatalf("child PID is wrong, expected %d got %d", expectedChildPid, pid)
+	}
+
+	_, err = createTree(nil)
+	if err == nil {
+		t.Fatalf("error is nil while it was expected when creating empty tree")
+	}
+}
+
+func TestTreeToString(t *testing.T) {
+	nodes := make([]*nodeProcess, 5)
+	i := 42
+	for j := range nodes {
+		nodes[j] = &nodeProcess{
+			event: &types.Event{
+				Command: fmt.Sprintf("foo-%d", i),
+				Pid:     i,
+			},
+			children: make([]*nodeProcess, 0),
+		}
+
+		i++
+	}
+
+	// Node 0 is the root and has two children: 1 and 2
+	nodes[0].children = append(nodes[0].children, nodes[1])
+	nodes[0].children = append(nodes[0].children, nodes[2])
+
+	// Node 2 has two children: 3 and 4 (which are leaf nodes)
+	nodes[2].children = append(nodes[2].children, nodes[3])
+	nodes[2].children = append(nodes[2].children, nodes[4])
+
+	i = 0
+	treeString := treeToString(nodes[0], "")
+	scanner := bufio.NewScanner(strings.NewReader(treeString))
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		expected := fmt.Sprintf("%s(%d)", nodes[i].event.Command, nodes[i].event.Pid)
```

Comment:
This test relies on the fact that the above appends will add the elements in order. And, as far as I know, it is not guaranteed, is it?

Reply:
Aaahh you are right, it is an array. Order is not guaranteed when we add elements to a map. Sorry, never mind.